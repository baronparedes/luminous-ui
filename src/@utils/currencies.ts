export function currencyFormat(num?: number) {
  num = num ? num : 0;
  return num.toString().replace(/(\d)(?=(\d{3})+(?!\d))/g, '$1,');
}

export function currencyFixedFormat(num?: number) {
  num = num ? num : 0;
  return num.toFixed(2).replace(/(\d)(?=(\d{3})+(?!\d))/g, '$1,');
}

export function roundOff(num?: number) {
  if (!num) return 0;
  return num === 0 ? 0 : Math.round(num * 100) / 100;
}
