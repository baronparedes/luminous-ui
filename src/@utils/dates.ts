import moment from 'moment';

import {Month, Period} from '../Api';

export function getCurrentDateFormatted() {
  return formatDate(new Date());
}

export function formatDate(date: Date | string, format = 'YYYY-MMM-DD') {
  return moment(date).format(format);
}

export function getCurrentDateTimeFormatted(date?: Date) {
  const now = date ?? new Date();
  return `${formatDate(now)} ${now.toLocaleTimeString()}`;
}

export function getCurrentMonthYear(): Period {
  const [year, month] = moment().format('YYYY MMM').split(' ');
  return {
    year: Number(year),
    month: month.toUpperCase() as Month,
  };
}

export function getCurrentMonthYearRelativeToCutoff(cutoffDay: number) {
  const currentDay = moment().day();
  const period = getCurrentMonthYear();
  if (currentDay > cutoffDay) {
    return addFromYearMonth(period.year, period.month, 1);
  }
  return period;
}

export function toMonthValue(value: Month) {
  return Number(moment().month(value).format('M'));
}

export function toMonthName(value: number) {
  return moment().month(value).format('MMM').toUpperCase() as Month;
}

export function toTransactionPeriod(year: number, month: Month) {
  const dateString = `${year}-${moment().month(month).format('MM')}-01`;
  return new Date(dateString);
}

export function toTransactionPeriodFromDate(date: Date): Period {
  const [year, month] = moment(date).format('YYYY MMM').split(' ');
  return {
    year: Number(year),
    month: month.toUpperCase() as Month,
  };
}

export function addFromYearMonth(year: number, month: Month, amount: number) {
  const result = moment().month(month).year(year).add(amount, 'month');
  const [x, y] = result.format('YYYY-MMM').split('-');
  return {
    year: Number(x),
    month: y.toUpperCase() as Month,
  };
}

export function subtractFromYearMonth(
  year: number,
  month: Month,
  amount: number
) {
  const result = moment().month(month).year(year).subtract(amount, 'month');
  const [x, y] = result.format('YYYY-MMM').split('-');
  return {
    year: Number(x),
    month: y.toUpperCase() as Month,
  };
}

export function getPastYears(count: number) {
  const {year: currentYear} = getCurrentMonthYear();
  const past = currentYear - count;
  const pastYears: number[] = [];
  for (let i = past + 1; i <= past + count; i++) {
    pastYears.push(i);
  }
  return pastYears;
}

export function getMonthsUpToCurrent() {
  const {month: currentMonth} = getCurrentMonthYear();
  const monthValue = toMonthValue(currentMonth);
  const months: Month[] = [];
  for (let i = 0; i <= monthValue - 1; i++) {
    months.push(toMonthName(i));
  }
  return months;
}
