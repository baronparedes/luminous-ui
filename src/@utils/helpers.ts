import {TransactionAttr} from '../Api';

export function generateNumberedSeries(n: number): number[] {
  return Array(n)
    .fill(0)
    .map((_, i) => i + 1);
}

export function sum(numbers: number[] | undefined) {
  if (!numbers || numbers.length === 0) return 0;
  return numbers.reduce((p, c) => Number(p) + Number(c));
}

export function sumTransactions(transactions: TransactionAttr[] | undefined) {
  if (!transactions || transactions.length === 0) return 0;
  const charged = transactions
    .filter(t => t.transactionType === 'charged')
    .map(t => t.amount);
  const collected = transactions
    .filter(t => t.transactionType === 'collected')
    .map(t => t.amount);
  return sum(charged) - sum(collected);
}
