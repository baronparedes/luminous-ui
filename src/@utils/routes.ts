const routes = {
  ROOT: '/',
  LOGIN: '/login',
  DASHBOARD: '/dashboard',
  PROPERTY: (id?: string | number) => `/property/${id}`,
  PURCHASE_ORDER: (id?: string | number) => `/purchase-order/${id}`,
  EXPENSE: '/expense',
  EXPENSE_REQUESTS: '/expense/requests',
  ADMIN: '/admin',
  ADMIN_PROFILES: '/admin/profiles',
  ADMIN_PROPERTIES: '/admin/properties',
  ADMIN_SETTINGS: '/admin/settings',
  ADMIN_BATCH_TRANSACTIONS: '/admin/batch-transactions',
  ADMIN_BATCH_PRINT_SOA: '/admin/batch-print-soa',
  ADMIN_UPLOAD_WATER_READING: '/admin/upload-water-reading',
};

export default routes;
