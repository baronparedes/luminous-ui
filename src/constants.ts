import {Month, Period, ProfileType, RecordStatus} from './Api';

export const STATUS_COLORS = {
  active: 'text-success',
  inactive: 'text-danger',
};

export const RECORD_STATUS: RecordStatus[] = ['active', 'inactive'];
export const PROFILE_TYPE: ProfileType[] = [
  'unit owner',
  'stakeholder',
  'admin',
];

export const SETTING_KEYS = {
  SOA_NOTES: 'SOA_NOTES',
  PR_NOTES: 'PR_NOTES',
  BILLING_CUTOFF_DAY: 'BILLING_CUTOFF_DAY',
};

export const DEFAULTS = {
  COMMUNITY_ID: 1,
  BILLING_CUTOFF_DAY: 30,
  WATER_CHARGE_ID: 6,
  COMMUNITY_EXPENSE: 'COMMUNITY EXPENSE',
  COMMUNITY_CHARGE_ID: 1,
};

export const VERBIAGE = {
  FILE_NAMES: {
    SOA_DOC_TITLE: (propertyCode: string | undefined, period: Period) =>
      `${propertyCode ?? 'Unit'} - SOA - ${period.year} ${period.month}`,
    PAYMENT_HISTORY_DOC_TITLE: (
      propertyCode: string | undefined,
      year: number
    ) => `${propertyCode ?? 'Unit'} - Payment History - ${year}`,
  },
};

export const MONTHS: Month[] = [
  'JAN',
  'FEB',
  'MAR',
  'APR',
  'MAY',
  'JUN',
  'JUL',
  'AUG',
  'SEP',
  'OCT',
  'NOV',
  'DEC',
];
