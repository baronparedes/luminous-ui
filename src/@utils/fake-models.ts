import faker from 'faker';

import {
  AuthProfile,
  AuthResult,
  ChargeAttr,
  ChargeType,
  PostingType,
  ProfileAttr,
  ProfileType,
  PropertyAccount,
  PropertyAssignmentAttr,
  PropertyAttr,
  RecordStatus,
  SettingAttr,
  TransactionAttr,
  TransactionType,
} from '../Api';
import {generateNumberedSeries} from './helpers';

export function generateFakeProfileAttr(type?: ProfileType): ProfileAttr {
  return {
    name: faker.name.findName(),
    username: faker.internet.userName(),
    password: faker.internet.password(),
    id: faker.datatype.number(),
    email: faker.internet.email(),
    mobileNumber: faker.phone.phoneNumber(),
    status: faker.random.arrayElement<RecordStatus>(['active', 'inactive']),
    type:
      type ??
      faker.random.arrayElement<ProfileType>([
        'unit owner',
        'stakeholder',
        'admin',
      ]),
  };
}

export function generateFakeProfile(type?: ProfileType): AuthProfile {
  return {
    name: faker.name.findName(),
    username: faker.random.words(1),
    id: faker.datatype.number(),
    email: faker.internet.email(),
    mobileNumber: faker.phone.phoneNumber(),
    status: faker.random.arrayElement<RecordStatus>(['active', 'inactive']),
    remarks: faker.random.words(),
    type:
      type ??
      faker.random.arrayElement<ProfileType>([
        'unit owner',
        'stakeholder',
        'admin',
      ]),
  };
}

export function generateFakeAuthResult(): AuthResult {
  return {
    profile: generateFakeProfile(),
    token: faker.random.alphaNumeric(100),
  };
}

export function generateFakeProperty(): PropertyAttr {
  return {
    id: faker.datatype.number(),
    code: faker.datatype.string(),
    floorArea: faker.datatype.number({precision: 2}),
    address: `${faker.address.cityName()} ${faker.address.country()}`,
    status: faker.random.arrayElement<RecordStatus>(['active', 'inactive']),
  };
}

export function generateFakePropertyAssignment(): PropertyAssignmentAttr {
  return {
    profileId: faker.datatype.number(),
    propertyId: faker.datatype.number(),
    profile: generateFakeProfileAttr(),
  };
}

export function generateFakeCharge(): ChargeAttr {
  return {
    chargeType: faker.random.arrayElement<ChargeType>([
      'amount',
      'percentage',
      'unit',
    ]),
    code: faker.random.words(),
    communityId: faker.datatype.number(),
    postingType: faker.random.arrayElement<PostingType>([
      'accrued',
      'manual',
      'monthly',
    ]),
    rate: faker.datatype.number(),
  };
}

export function generateFakeTransaction(): TransactionAttr {
  const charge = generateFakeCharge();
  return {
    amount: faker.datatype.number(),
    chargeId: faker.datatype.number(),
    propertyId: faker.datatype.number(),
    transactionPeriod: faker.date.recent().toDateString(),
    transactionType: faker.random.arrayElement<TransactionType>([
      'charged',
      'collected',
    ]),
    id: faker.datatype.number(),
    charge,
  };
}

export function generateFakePropertyAccount(
  transactionCount = 2
): PropertyAccount {
  const property = generateFakeProperty();
  const transactions = generateNumberedSeries(transactionCount).map(() =>
    generateFakeTransaction()
  );
  return {
    propertyId: Number(property.id),
    property,
    balance: faker.datatype.number(),
    transactions,
  };
}

export const generateFakeSetting = (): SettingAttr => {
  return {
    key: faker.random.word(),
    value: faker.random.words(),
  };
};
