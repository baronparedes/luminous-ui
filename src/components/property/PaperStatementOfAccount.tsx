import React from 'react';

import {SETTING_KEYS} from '../../constants';
import {useRootState} from '../../store';
import {PrintPaper} from '../@ui/PaperPdf';
import SOA from '../@ui/SOA';

type Props = React.ComponentProps<typeof SOA>;

const PaperStatementOfAccount = React.forwardRef<HTMLDivElement, Props>(
  ({...soaProps}, ref) => {
    const notes = useRootState(state =>
      state.setting.values.find(v => v.key === SETTING_KEYS.SOA_NOTES)
    );
    return (
      <div className="d-none">
        <PrintPaper ref={ref}>
          <SOA {...soaProps} notes={notes} />
        </PrintPaper>
      </div>
    );
  }
);

export default PaperStatementOfAccount;
