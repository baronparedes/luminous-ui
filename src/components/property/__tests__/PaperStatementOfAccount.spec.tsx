import {toTransactionPeriod} from '../../../@utils/dates';
import {renderWithProvider} from '../../../@utils/test-renderers';
import {PropertyAccount, SettingAttr} from '../../../Api';
import {SETTING_KEYS} from '../../../constants';
import {settingActions} from '../../../store/reducers/setting.reducer';
import PaperStatementOfAccount from '../PaperStatementOfAccount';

describe('PaperStatementOfAccount', () => {
  const propertyAccount: PropertyAccount = {
    balance: 2000,
    propertyId: 1,
    property: {
      id: 1,
      code: 'G-111',
      address: 'Pasig City Philippines',
      floorArea: 33.1,
      status: 'active',
    },
    transactions: [
      {
        amount: 1000,
        chargeId: 1,
        propertyId: 1,
        transactionPeriod: toTransactionPeriod(2021, 'SEP').toString(),
        transactionType: 'charged',
        charge: {
          code: 'DUES',
          rate: 50,
          communityId: 1,
          chargeType: 'unit',
          postingType: 'monthly',
        },
      },
      {
        amount: 500,
        chargeId: 1,
        propertyId: 1,
        transactionPeriod: toTransactionPeriod(2021, 'SEP').toString(),
        transactionType: 'collected',
        paymentDetailId: 1,
        charge: {
          code: 'DUES',
          rate: 50,
          communityId: 1,
          chargeType: 'unit',
          postingType: 'monthly',
        },
      },
      {
        amount: 300,
        chargeId: 1,
        propertyId: 1,
        transactionPeriod: toTransactionPeriod(2021, 'SEP').toString(),
        transactionType: 'collected',
        paymentDetailId: 2,
        charge: {
          code: 'DUES',
          rate: 50,
          communityId: 1,
          chargeType: 'unit',
          postingType: 'monthly',
        },
      },
    ],
    paymentDetails: [
      {
        id: 1,
        collectedBy: 1,
        orNumber: '1234-1234',
        paymentType: 'cash',
      },
      {
        id: 2,
        collectedBy: 1,
        orNumber: '1234-1234',
        paymentType: 'check',
        checkIssuingBank: 'BDO',
        checkNumber: '111-111-000',
        checkPostingDate: '2021-09-09',
      },
    ],
    assignedProfiles: [
      {
        email: 'JohnDoe@email.com',
        password: '',
        status: 'active',
        type: 'unit owner',
        username: 'johndoe',
        name: 'John Doe',
      },
    ],
  };

  const settings: SettingAttr[] = [
    {
      key: SETTING_KEYS.SOA_NOTES,
      value: '<h1>SOA Notes</h1>',
    },
  ];

  it('should render', () => {
    const {asFragment} = renderWithProvider(
      <PaperStatementOfAccount
        year={2021}
        month={'SEP'}
        propertyAccount={propertyAccount}
      />,
      store => store.dispatch(settingActions.init(settings))
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
