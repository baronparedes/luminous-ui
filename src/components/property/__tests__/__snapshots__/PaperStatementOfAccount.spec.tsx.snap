// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaperStatementOfAccount should render 1`] = `
<DocumentFragment>
  <div
    class="d-none"
  >
    <div>
      <div
        class="sc-bdfBQB dxnWgn"
      >
        <div
          class="sc-bdfBQB dxnWgn"
        >
          <div
            class="text-center"
          >
            <h1
              class="brand"
            >
              Luminous
            </h1>
            <h4>
              STATEMENT OF ACCOUNT
            </h4>
          </div>
        </div>
        <div
          class="sc-bdfBQB dxnWgn pt-3"
        >
          <div
            class="sc-gsTEea bjzFuM"
          >
            <strong>
              Unit Number: 
            </strong>
            G-111
          </div>
          <div
            class="sc-gsTEea bjzFuM"
          >
            <strong>
              Unit owner: 
            </strong>
            John Doe
          </div>
          <small>
            <div
              class="sc-gsTEea bjzFuM"
            >
              <strong>
                Address: 
              </strong>
              Pasig City Philippines
            </div>
          </small>
          <small>
            <div
              class="sc-gsTEea bjzFuM"
            >
              <strong>
                Floor area: 
              </strong>
              33.1
            </div>
          </small>
          <hr />
        </div>
        <div
          class="sc-bdfBQB dxnWgn pt-3"
        >
          <div
            class="sc-gsTEea bjzFuM"
          >
            <strong>
              Billing Period: 
            </strong>
            SEP 2021
          </div>
          <hr />
          <table
            class="w-100"
          >
            <thead>
              <tr>
                <th>
                  Charge Description
                </th>
                <th>
                  Rate
                </th>
                <th
                  class="float-right"
                >
                  Amount
                </th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td>
                  DUES
                </td>
                <td>
                  50
                </td>
                <td
                  class="float-right"
                >
                  <strong>
                    <span
                      class="currency"
                    >
                      1,000
                    </span>
                  </strong>
                </td>
              </tr>
            </tbody>
          </table>
          <hr />
        </div>
        <div
          class="sc-bdfBQB dxnWgn"
        >
          <small>
            <div
              class="sc-gsTEea bjzFuM"
            >
              <strong
                class="text-muted"
              >
                Received Payments
              </strong>
            </div>
          </small>
          <div
            class="container"
          >
            <div
              class="row"
            >
              <div
                class="p-0 m-0 col"
              >
                <div>
                  <div
                    class="d-inline pr-2"
                  >
                    <span
                      class="text-muted pr-2"
                    >
                      OR#
                    </span>
                    1234-1234
                  </div>
                  <div
                    class="d-inline pr-2"
                  >
                    <span
                      class="text-muted pr-2"
                    >
                      received
                    </span>
                    cash
                  </div>
                  <div
                    class="d-inline pr-2"
                  >
                    <span
                      class="text-muted pr-2"
                    >
                      with an amount of
                    </span>
                    <span
                      class="currency"
                    >
                      500
                    </span>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="row"
            >
              <div
                class="p-0 m-0 col"
              >
                <div>
                  <div
                    class="d-inline pr-2"
                  >
                    <span
                      class="text-muted pr-2"
                    >
                      OR#
                    </span>
                    1234-1234
                  </div>
                  <div
                    class="d-inline pr-2"
                  >
                    <span
                      class="text-muted pr-2"
                    >
                      received
                    </span>
                    check
                  </div>
                  <div
                    class="d-inline pr-2"
                  >
                    <span
                      class="text-muted pr-2"
                    >
                      with an amount of
                    </span>
                    <span
                      class="currency"
                    >
                      300
                    </span>
                  </div>
                  <div
                    class="text-muted"
                  >
                    <small
                      class="pr-3"
                    >
                      <span>
                        BDO
                      </span>
                    </small>
                    <small
                      class="pr-3"
                    >
                      <span>
                        111-111-000
                      </span>
                    </small>
                    <small>
                      <span>
                        2021-09-09
                      </span>
                    </small>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <hr />
        </div>
        <div
          class="sc-bdfBQB dxnWgn pt-1"
        >
          <small>
            <div
              class="sc-gsTEea bjzFuM"
            >
              <strong
                class="text-muted"
              >
                Current Balance
              </strong>
              <strong
                class="float-right"
              >
                <span
                  class="currency"
                >
                  1,000
                </span>
              </strong>
            </div>
          </small>
          <small>
            <div
              class="sc-gsTEea bjzFuM"
            >
              <strong
                class="text-muted"
              >
                Previous Balance
              </strong>
              <strong
                class="float-right"
              >
                <span
                  class="currency"
                >
                  1,800
                </span>
              </strong>
            </div>
          </small>
          <small>
            <div
              class="sc-gsTEea bjzFuM"
            >
              <strong
                class="text-muted"
              >
                Less Payments
              </strong>
              <strong
                class="float-right"
              >
                <span
                  class="currency"
                >
                  800
                </span>
              </strong>
            </div>
          </small>
          <hr />
          <div
            class="sc-gsTEea bjzFuM"
          >
            <strong
              class="text-muted"
            >
              Total Amount Due
            </strong>
            <strong
              class="float-right"
            >
              <span
                class="currency"
              >
                2,000
              </span>
            </strong>
          </div>
        </div>
        <div
          class="sc-bdfBQB dxnWgn pt-3"
        >
          <hr />
          <div>
            <h1>
              SOA Notes
            </h1>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
